README.md
# Welcome to ogs5py

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.2546767.svg)](https://doi.org/10.5281/zenodo.2546767)
[![PyPI version](https://badge.fury.io/py/ogs5py.svg)](https://badge.fury.io/py/ogs5py)
[![Continuous Integration](https://github.com/GeoStat-Framework/ogs5py/actions/workflows/main.yml/badge.svg)](https://github.com/GeoStat-Framework/ogs5py/actions/workflows/main.yml)
[![Coverage Status](https://coveralls.io/repos/github/GeoStat-Framework/ogs5py/badge.svg?branch=main)](https://coveralls.io/github/GeoStat-Framework/ogs5py?branch=main)
[![Documentation Status](https://readthedocs.org/projects/ogs5py/badge/?version=stable)](https://geostat-framework.readthedocs.io/projects/ogs5py/en/stable/?badge=stable)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black)

<p align="center">
<img src="https://raw.githubusercontent.com/GeoStat-Framework/ogs5py/main/docs/source/pics/OGS.png" alt="ogs5py-LOGO" width="251px"/>
</p>

## Purpose

ogs5py is A python-API for the [OpenGeoSys 5][ogs5_link] scientific modeling package.


## Installation

You can install the latest version with the following command:

    pip install ogs5py

Or with conda:

    conda install ogs5py


## Citation

If you are using ogs5py in your publication please cite our paper:

> Müller, S., Zech, A. and Heße, F.:
> ogs5py: A Python-API for the OpenGeoSys 5 Scientific Modeling Package.
> Groundwater, 59: 117-122. https://doi.org/10.1111/gwat.13017, 2021.

You can cite the Zenodo code publication of ogs5py by:

> Sebastian Müller. GeoStat-Framework/ogs5py. Zenodo. https://doi.org/10.5281/zenodo.2546767

If you want to cite a specific version, have a look at the [Zenodo site](https://doi.org/10.5281/zenodo.2546767).


## Documentation for ogs5py

You can find the documentation under [ogs5py.readthedocs.io][doc_link].


### Further Information

- General homepage: https://www.opengeosys.org/ogs-5
- OGS5 Repository: https://github.com/ufz/ogs5
- Keyword documentation: https://ogs5-keywords.netlify.com
- OGS5 Benchmarks: https://github.com/ufz/ogs5-benchmarks
- ogs5py Benchmarks: https://github.com/GeoStat-Framework/ogs5py_benchmarks


### Tutorials and Examples

In the following a simple transient pumping test is simulated on a radial symmetric mesh.
The point output at the observation well is plotted afterwards.
For more details on this example, see: [Tutorial 1][tut1_link]

```python
from ogs5py import OGS, specialrange, generate_time
from matplotlib import pyplot as plt

# discretization and parameters
time = specialrange(0, 3600, 50, typ="cub")
rad = specialrange(0, 1000, 100, typ="cub")
obs = rad[21]
angles = 32
storage = 1e-3
transmissivity = 1e-4
rate = -1e-3
# model setup
model = OGS(task_root="pump_test", task_id="model")
# generate a radial mesh and geometry ("boundary" polyline)
model.msh.generate("radial", dim=2, rad=rad, angles=angles)
model.gli.generate("radial", dim=2, rad_out=rad[-1], angles=angles)
model.gli.add_points([0.0, 0.0, 0.0], "pwell")
model.gli.add_points([obs, 0.0, 0.0], "owell")
model.bc.add_block(  # boundary condition
    PCS_TYPE="GROUNDWATER_FLOW",
    PRIMARY_VARIABLE="HEAD",
    GEO_TYPE=["POLYLINE", "boundary"],
    DIS_TYPE=["CONSTANT", 0.0],
)
model.st.add_block(  # source term
    PCS_TYPE="GROUNDWATER_FLOW",
    PRIMARY_VARIABLE="HEAD",
    GEO_TYPE=["POINT", "pwell"],
    DIS_TYPE=["CONSTANT_NEUMANN", rate],
)
model.ic.add_block(  # initial condition
    PCS_TYPE="GROUNDWATER_FLOW",
    PRIMARY_VARIABLE="HEAD",
    GEO_TYPE="DOMAIN",
    DIS_TYPE=["CONSTANT", 0.0],
)
model.mmp.add_block(  # medium properties
    GEOMETRY_DIMENSION=2,
    STORAGE=[1, storage],
    PERMEABILITY_TENSOR=["ISOTROPIC", transmissivity],
)
model.num.add_block(  # numerical solver
    PCS_TYPE="GROUNDWATER_FLOW",
    LINEAR_SOLVER=[2, 5, 1e-14, 1000, 1.0, 100, 4],
)
model.out.add_block(  # point observation
    PCS_TYPE="GROUNDWATER_FLOW",
    NOD_VALUES="HEAD",
    GEO_TYPE=["POINT", "owell"],
    DAT_TYPE="TECPLOT",
)
model.pcs.add_block(  # set the process type
    PCS_TYPE="GROUNDWATER_FLOW", NUM_TYPE="NEW"
)
model.tim.add_block(  # set the timesteps
    PCS_TYPE="GROUNDWATER_FLOW",
    **generate_time(time)
)
model.write_input()
model.run_model()
```

<p align="center">
<img src="https://raw.githubusercontent.com/GeoStat-Framework/ogs5py/main/docs/source/pics/01_pump_test_drawdown.png" alt="Drawdown" width="600px"/>
</p>


### OGS5 executable

To obtain an OGS5 executable, ``ogs5py`` brings a download routine:

```python
from ogs5py import download_ogs
download_ogs()
```

Then a executable is stored in the ogs5py config path and will be called
when a model is run.

You can pass a ``version`` statement to the ``download_ogs`` routine, to
obtain a specific version (5.7, 5.7.1 (win only) and 5.8).
For OGS 5.7 there are executables for Windows/Linux and MacOS.
For "5.8" there are no MacOS pre-builds.

If you have compiled your own OGS5 version, you can add your executable
to the ogs5py config path with:

```python
from ogs5py import add_exe
add_exe("path/to/your/ogs/exe")
```

Otherwise you need to specify the path to the executable within the run command:

```python
model.run_model(ogs_exe="path/to/ogs")
```


## Requirements:

- [NumPy >= 1.14.5](https://www.numpy.org)
- [Pandas >= 0.23.2](https://pandas.pydata.org/)
- [meshio >= 4](https://github.com/nschloe/meshio)
- [lxml >= 4](https://github.com/lxml/lxml)
- [pexpect >= 4](https://github.com/pexpect/pexpect)
- [vtk >= 9](https://vtk.org/)

## Contact

You can contact us via <info@geostat-framework.org>.


## License

[MIT][gpl_link] © 2018-2023 (inspired by Falk Hesse and Miao Jing)

This project is based on [OGSPY][ogspy_link].

[ogspy_link]: https://github.com/fhesze/OGSPY
[gpl_link]: https://github.com/GeoStat-Framework/ogs5py/blob/main/LICENSE
[ogs5_link]: https://www.opengeosys.org/ogs-5/
[doc_link]: https://ogs5py.readthedocs.io/
[tut1_link]: https://geostat-framework.readthedocs.io/projects/ogs5py/en/stable/tutorial_01_pump.html


CHANGELOG.md
# Changelog

All notable changes to **ogs5py** will be documented in this file.


## [1.3.0] - 2023-04

See [#18](https://github.com/GeoStat-Framework/ogs5py/pull/18)

### Enhancements
- move to `src/` base package structure
- use [hatchling](https://pypi.org/project/hatchling/) as build backend
- drop py36 support
- added archive support
- simplify documentation

### Bugfixes
- remove usage of deprecated `np.int`


## [1.2.2] - 2022-05-25

### Bugfixes
* `MSH.load` now uses `engine="python"` as fallback in pandas to successfully read ogs meshes in some corner cases [#16](https://github.com/GeoStat-Framework/ogs5py/pull/16)
* removed redundant `from io import open` which were there for py2 compatibility [#16](https://github.com/GeoStat-Framework/ogs5py/pull/16)


## [1.2.1] - 2022-05-15

### Enhancements
* `MSH.import_mesh` can handle `meshio.Mesh` as input now [#13](https://github.com/GeoStat-Framework/ogs5py/pull/13)

### Changes
* `pygmsh` support was removed. You can't use pygmsh Geometry objects to generate meshes anymore. Please generate beforehand and import the generated mesh. Other generators are using `gmsh` directly now. [#13](https://github.com/GeoStat-Framework/ogs5py/pull/13)


## [1.2.0] - 2022-05-15

### Enhancements
* move to a `pyproject.toml` based installation: [d5ea756](https://github.com/GeoStat-Framework/ogs5py/commit/d5ea756629f3aaad6f4f8e197926f932677a6658)
* move from `develop`/`master` branches to a single `main` branch
* use GitHub Actions for CI: [b6811ce](https://github.com/GeoStat-Framework/ogs5py/commit/b6811ce3c39fa0eea849aac146384032cda10a37)
* use f-strings where possible [#11](https://github.com/GeoStat-Framework/ogs5py/pull/11)
* simplified documentation [#11](https://github.com/GeoStat-Framework/ogs5py/pull/11)
* added changelog to documentation [#11](https://github.com/GeoStat-Framework/ogs5py/pull/11)
* added citation file and paper reference [#11](https://github.com/GeoStat-Framework/ogs5py/pull/11)
* use Python 3 style classes [#11](https://github.com/GeoStat-Framework/ogs5py/pull/11)

### Changes
* `downlaod_ogs` only supports version "5.7", "5.7.1" and "5.8" since the CI for OGS5 was shut down: [8b1cc91](https://github.com/GeoStat-Framework/ogs5py/commit/8b1cc91e5e88ac65288bc30b9c44fda781b70710)

### Bugfixes
* make `downlaod_ogs` work again [8b1cc91](https://github.com/GeoStat-Framework/ogs5py/commit/8b1cc91e5e88ac65288bc30b9c44fda781b70710)
* documentation fix in `GLI.add_polyline` [#7](https://github.com/GeoStat-Framework/ogs5py/issues/7)
* require `pygmsh<7` for now [#11](https://github.com/GeoStat-Framework/ogs5py/pull/11)


## [1.1.1] - 2020-04-02

### Bugfixes
* check if `__version__` is present (only if installed)


## [1.1.0] - 2020-03-22

### Bugfixes
* meshio 4 was not compatible
* fixed integer type in exporting meshes with element/material IDs
* better check for OGS5 success on Windows

### Changes
* drop py2.7 support


## [1.0.5] - 2019-11-18

### Bugfixes
* ``MSH.set_material_id``: better handling of non-int single values: [f34d2e5](https://github.com/GeoStat-Framework/ogs5py/commit/f34d2e510e7bb340abc19150270b956c407d7ff6)
* ``MSH.show``: better handling of material IDs: [26b4610](https://github.com/GeoStat-Framework/ogs5py/commit/26b46108a14f4fe95e0b7424bb51e4482c75eae9)
* ``GLI.add_polyline``: Adding polyline by point-names was not possible: [17dd199](https://github.com/GeoStat-Framework/ogs5py/commit/17dd19944f49666f00d09fe2d21e88345439cf4d)

### Additions
* better integration of pygmsh: [570afd9](https://github.com/GeoStat-Framework/ogs5py/commit/570afd9d415c9fa402788c82f4864b7e6ddbf5d9)
* new functions ``specialrange`` and ``generate_time``: [e5f3aba](https://github.com/GeoStat-Framework/ogs5py/commit/e5f3aba542c41116e0fe0415e0eea6e689a7ccee)
* updated examples


## [1.0.4] - 2019-09-10

### Bugfixes
* ogs5py was not usable offline: [0f98c32](https://github.com/GeoStat-Framework/ogs5py/commit/0f98c32213d02b4b46455a5392b3c1c6fc021228)
* ``add_exe`` was not recognizing operation system: [89b07e5](https://github.com/GeoStat-Framework/ogs5py/commit/89b07e51df4b4f31b8c42c56e4ad24063c57f845)

### Additions
* new sub-keywords for OUT (added to OGS5 in Aug 19) when using TECPLOT (TECPLOT_ELEMENT_OUTPUT_CELL_CENTERED, TECPLOT_ZONES_FOR_MG): [ebcb22a](https://github.com/GeoStat-Framework/ogs5py/commit/ebcb22acb9afff83cc4593425d5dcde1bf97b9dc)

### Changes
* RFR Class was refactored to allow multiple variables: [3c1c445](https://github.com/GeoStat-Framework/ogs5py/commit/3c1c44533a9640ac1f2c173c321a029141f7aa2b)


## [1.0.3] - 2019-08-23

### Bugfixes
* ``MSH.show`` TempFile was not working on Windows: [c0d0960](https://github.com/GeoStat-Framework/ogs5py/commit/c0d0960b02c73237120030d6b1b23082dea085bc)


## [1.0.2] - 2019-08-22

### Bugfixes
* Don't fix QT_API for MAYAVI and use vtk for export: [33398ad](https://github.com/GeoStat-Framework/ogs5py/commit/33398adb4f3e1750e037434b8096c8eae7f0f419)
* ``PopenSpawn`` has no close attribute on Windows: [12f05d6](https://github.com/GeoStat-Framework/ogs5py/commit/12f05d6bcc5b5992f0d6aeb6feee09a4ca2f11b9)


## [1.0.1] - 2019-08-22

### Bugfixes
* ``download_ogs(version="latest" build="PETSC")`` was not working: [552503b](https://github.com/GeoStat-Framework/ogs5py/commit/552503b030634465aa324ebf3401fed8805c856b)


## [1.0.0] - 2019-08-22

### Bugfixes
* ``GLI.add_polyline`` now allows integer coordinates for points: [bf5d684](https://github.com/GeoStat-Framework/ogs5py/commit/bf5d684a2c01f2b6eb4e13098cfc77711c05d35d)
* ``MSH.centroids`` are now calculated as center of mass instead of center of element nodes: [b0708a6](https://github.com/GeoStat-Framework/ogs5py/commit/b0708a69bd290d613d663385193a3a9ff69ee625)
* ``MSH.show`` was not working: [6a0489b](https://github.com/GeoStat-Framework/ogs5py/commit/6a0489bc1675b909946e1a870459fc7f7ddf7629)

ogs5py/tools/types.py
# -*- coding: utf-8 -*-
"""
type definitions for ogs5

GLI related Constants
^^^^^^^^^^^^^^^^^^^^^

.. autosummary::
   EMPTY_GLI
   GLI_KEYS
   GLI_KEY_LIST
   EMPTY_PLY
   PLY_KEYS
   PLY_KEY_LIST
   PLY_TYPES
   EMPTY_SRF
   SRF_KEYS
   SRF_KEY_LIST
   SRF_TYPES
   EMPTY_VOL
   VOL_KEYS
   VOL_KEY_LIST
   VOL_TYPES

MSH related Constants
^^^^^^^^^^^^^^^^^^^^^

.. autosummary::
   EMPTY_MSH
   MESH_KEYS
   MESH_DATA_KEYS
   ELEM_1D
   ELEM_2D
   ELEM_3D
   ELEM_DIM
   ELEM_NAMES
   ELEM_TYP
   ELEM_TYP1D
   ELEM_TYP2D
   ELEM_TYP3D
   VTK_TYP
   MESHIO_NAMES
   NODE_NO

General Constants
^^^^^^^^^^^^^^^^^

.. autosummary::
   STRTYPE
   PCS_TYP
   PCS_EXT
   PRIM_VAR
   PRIM_VAR_BY_PCS
   OGS_EXT
   MULTI_FILES

"""
import numpy as np

STRTYPE = str
"""type: base string type"""

# keylists for the gli entries and templates for entries
EMPTY_GLI = {
    "points": None,
    "point_names": None,
    "point_md": None,
    "polylines": [],
    "surfaces": [],
    "volumes": [],
}
"""dict: empty gli dict"""

GLI_KEY_LIST = ["#POINTS", "#POLYLINE", "#SURFACE", "#VOLUME", "#STOP"]
"""list: gli main keys"""

# https://github.com/ufz/ogs5/blob/e704a791391a233084c3d74e1335f50206c5eb76/GEO/geo_ply.cpp#L577
PLY_KEY_LIST = [
    "ID",  # int
    "NAME",  # str
    "TYPE",  # int
    "EPSILON",  # float
    "MAT_GROUP",  # int
    "POINTS",  # list
    "POINT_VECTOR",  # str
]
"""list: gli polyline keys"""

PLY_TYPES = [int, str, int, float, int, list, str]
"""list: gli polyline key types"""

# https://github.com/ufz/ogs5/blob/e704a791391a233084c3d74e1335f50206c5eb76/GEO/geo_sfc.cpp#L996
SRF_KEY_LIST = [
    "ID",  # int
    "NAME",  # str
    "EPSILON",  # float
    "TYPE",  # int
    "TIN",  # str
    "MAT_GROUP",  # int
    "POLYLINES",  # list
]
"""list: gli surface keys"""

SRF_TYPES = [int, str, float, int, str, int, list]
"""list: gli surface key types"""

# https://github.com/ufz/ogs5/blob/e704a791391a233084c3d74e1335f50206c5eb76/GEO/geo_vol.cpp#L130
VOL_KEY_LIST = [
    "NAME",  # str
    "TYPE",  # str
    "SURFACES",  # list
    "MAT_GROUP",  # str
    "LAYER",  # int
]
"""list: gli volume keys"""

VOL_TYPES = [str, str, list, str, int]
"""list: gli volume key types"""

EMPTY_PLY = {}
"""dict: empty ogs gli polyline dict"""

for key in PLY_KEY_LIST:
    EMPTY_PLY[key] = None
EMPTY_SRF = {}
"""dict: empty ogs gli surface dict"""

for key in SRF_KEY_LIST:
    EMPTY_SRF[key] = None
EMPTY_VOL = {}
"""dict: empty ogs gli volume dict"""

for key in VOL_KEY_LIST:
    EMPTY_VOL[key] = None
# keys for the gli-dict
GLI_KEYS = {
    "points",
    "point_names",
    "point_md",
    "polylines",
    "surfaces",
    "volumes",
}
"""set: ogs gli dict keys"""

PLY_KEYS = {
    "ID",
    "NAME",
    "POINTS",
    "EPSILON",
    "TYPE",
    "MAT_GROUP",
    "POINT_VECTOR",
}
"""set: ogs gli polyline keys"""

SRF_KEYS = {"ID", "NAME", "POLYLINES", "EPSILON", "TYPE", "MAT_GROUP", "TIN"}
"""set: ogs gli surface keys"""

VOL_KEYS = {"NAME", "SURFACES", "TYPE", "MAT_GROUP", "LAYER"}
"""set: ogs gli volume keys"""

# names sorted by dimensionality
ELEM_1D = ["line"]
"""set: ogs element names"""

ELEM_2D = ["tri", "quad"]
"""set: ogs element names"""

ELEM_3D = ["tet", "pyra", "pris", "hex"]
"""set: ogs element names"""

# names sorted by dim
ELEM_DIM = [ELEM_1D, ELEM_2D, ELEM_3D]
"""set: ogs element names"""

# all names for element types in ogs 5
ELEM_NAMES = ELEM_1D + ELEM_2D + ELEM_3D
"""list: ogs element names"""

# keys for the mesh-dict
MESH_KEYS = {"mesh_data", "nodes", "elements", "material_id", "element_id"}
"""set: ogs mesh dict-keys"""

MESH_DATA_KEYS = {
    "AXISYMMETRY",
    "CROSS_SECTION",
    "PCS_TYPE",
    "GEO_TYPE",
    "GEO_NAME",
    "LAYER",
}
"""set: ogs mesh data keys"""

ELEMENT_KEYS = set(ELEM_NAMES)
"""set: ogs element names"""

EMPTY_MSH = {
    "mesh_data": {},
    "nodes": np.empty((0, 3)),
    "elements": {},
    "element_id": {},
    "material_id": {},
}
"""dict: empty mesh dict"""

# coresponding names for types in meshio
MESHIO_NAMES = [
    "line",
    "triangle",
    "quad",
    "tetra",
    "pyramid",
    "wedge",
    "hexahedron",
]
"""list: coresponding element names in meshio"""

# number encoding for element types (obsolete)
ELEM_TYP = {
    0: "line",
    1: "tri",
    2: "quad",
    3: "tet",
    4: "pyra",
    5: "pris",
    6: "hex",
    "line": 0,
    "tri": 1,
    "quad": 2,
    "tet": 3,
    "pyra": 4,
    "pris": 5,
    "hex": 6,
}
"""dict: type code per element name"""

# number encoding sorted by dimensionality
ELEM_TYP1D = {0: "line", "line": 0}
"""dict: type code per element name"""

ELEM_TYP2D = {1: "tri", 2: "quad", "tri": 1, "quad": 2}
"""dict: type code per element name"""

ELEM_TYP3D = {
    3: "tet",
    4: "pyra",
    5: "pris",
    6: "hex",
    "tet": 3,
    "pyra": 4,
    "pris": 5,
    "hex": 6,
}
"""dict: type code per element name"""

# coresponding vtk-types by their number encoding
VTK_TYP = {
    3: "line",  # vtk.VTK_LINE == 3
    5: "tri",  # vtk.VTK_TRIANGLE == 5
    9: "quad",  # vtk.VTK_QUAD == 9
    10: "tet",  # vtk.VTK_TETRA == 10
    14: "pyra",  # vtk.VTK_PYRAMID == 14
    13: "pris",  # vtk.VTK_WEDGE == 13
    12: "hex",  # vtk.VTK_HEXAHEDRON == 12
    "line": 3,
    "tri": 5,
    "quad": 9,
    "tet": 10,
    "pyra": 14,
    "pris": 13,
    "hex": 12,
}
"""dict: vtk type codes per element name"""

# number of nodes per element-type (sorted by name and number-encoding)
NODE_NO = {
    0: 2,
    1: 3,
    2: 4,
    3: 4,
    4: 5,
    5: 6,
    6: 8,
    "line": 2,
    "tri": 3,
    "quad": 4,
    "tet": 4,
    "pyra": 5,
    "pris": 6,
    "hex": 8,
}
"""dict: Node numbers per element name"""

# all pcs types supported by OGS5
# https://ogs5-keywords.netlify.com/ogs/wiki/public/doc-auto/by_ext/pcs/s_pcs_type
PCS_TYP = [
    "",  # special case for no specified pcs in "*.out"
    "GROUNDWATER_FLOW",  # HEAD
    "LIQUID_FLOW",  # PRESSURE1
    "RICHARDS_FLOW",  # PRESSURE1
    "AIR_FLOW",  # PRESSURE1 TEMPERATURE1
    "MULTI_PHASE_FLOW",  # PRESSURE1, PRESSURE2
    "PS_GLOBAL",  # PRESSURE1, SATURATION2
    "HEAT_TRANSPORT",  # TEMPERATURE1
    "DEFORMATION",  # DISPLACEMENT_X1, -_Y1, -_Z1
    "MASS_TRANSPORT",  # varying (#COMP_PROP*.mcp 7 CONCENTRATION)
    "OVERLAND_FLOW",  # HEAD
    "FLUID_MOMENTUM",  # VELOCITY1_X, VELOCITY1_Y, VELOCITY1_Z
    "RANDOM_WALK",  # with particles in *.pct file
    "NO_PCS",  # ...used in Benchmarks
    "TNEQ",  # ...used in Benchmarks
    "TES",  # ...used in Benchmarks
    "DEFORMATION_SINGLEFLOW_MONO",  # ...used in Benchmarks
    "MULTI_COMPONENTIAL_FLOW",  # ...used in Benchmarks
]
"""list: PCS types"""

# file extensions by pcs type (and the unspecified case "")
PCS_EXT = [""] + ["_" + pcs for pcs in PCS_TYP[1:]]
"""list: PCS file extensions with _"""

# all PRIMARY_VARIABLE types supported by OGS5 (sorted by PCS_TYP)
PRIM_VAR = [
    [""],
    ["HEAD"],
    ["PRESSURE1"],
    ["PRESSURE1"],
    ["PRESSURE1", "TEMPERATURE1"],
    ["PRESSURE1", "PRESSURE2"],
    ["PRESSURE1, SATURATION2"],
    ["TEMPERATURE1"],
    ["DISPLACEMENT_X1", "DISPLACEMENT_Y1", "DISPLACEMENT_Z1"],
    [""],  # varying (#COMP_PROP*.mcp 7 CONCENTRATION)
    ["HEAD"],
    ["VELOCITY1_X", "VELOCITY1_Y", "VELOCITY1_Z"],
    [""],  # with particles in *.pct file
    [""],
    [""],
    [""],
    [""],
    [""],
]
"""list: primary variables"""

PRIM_VAR_BY_PCS = {}
"""dict: primary variables per PCS"""

for i, pcs in enumerate(PCS_TYP):
    PRIM_VAR_BY_PCS[pcs] = PRIM_VAR[i]
# file extensions of ogs5 input files (without mpd, gli_ext, rfr files)
OGS_EXT = [
    ".msh",  # Mesh
    ".gli",  # Geometry
    ".ddc",  # MPI domain decomposition
    ".pcs",  # Process settings
    ".rfd",  # definition of time-curves for variing BCs or STs
    ".cct",  # Communication Table
    ".fct",  # Function
    ".bc",  # Boundary Condition
    ".ic",  # Initial Condition
    ".st",  # Source Term
    ".mmp",  # Medium Properties
    ".msp",  # Solid Properties
    ".mfp",  # Fluid Properties
    ".mcp",  # reactive components for modelling chemical processes
    ".gem",  # geochemical thermodynamic modeling coupling
    ".krc",  # Kinetric Reaction
    ".pqc",  # Phreqqc coupling
    ".rei",  # Reaction Interface
    ".pct",  # Particle Definition for RANDOM_WALK
    ".num",  # Settings for the numerical solver
    ".tim",  # Time settings
    ".out",  # Output Settings
]
"""list: all ogs file extensions"""

MULTI_FILES = ["mpd", "gli_ext", "rfr", "gem_init", "asc"]
"""list: all ogs files that can occure multiple times"""
